projects/healthPingValidation.kp/knowledge.md

```diff
@@ -0,0 +1,550 @@
+    os: {u'version': u'10.0', u'name': u'WINNT'} reason: shutdown
+    
+    
+
+
+# Nightly 56 & 57. Compute pings count per day
+This include showing diagrams and printings stats
+
+
+```python
+from datetime import datetime
+
+def pingsCountPerDay(pings):
+    return pings.map(lambda ping: ping["meta/submissionDate"]).countByValue()
+
+resultDictionary = pingsCountPerDay(cachedData)
+```
+
+```python
+import matplotlib.dates as mdates
+def plotlistofTuples(listOfTuples, title="", inColor='blue'):
+    keys = [t[0] for t in listOfTuples]
+    values = [t[1] for t in listOfTuples]
+
+    plt.figure(1)
+    fig = plt.gcf()
+    fig.set_size_inches(15, 7)
+
+    plt.title(title)
+    plt.gca().xaxis.set_major_formatter(mdates.DateFormatter('%m/%d/%Y'))
+    plt.bar(range(len(listOfTuples)), values, align='center', color=inColor)
+    plt.xticks(range(len(listOfTuples)), keys, rotation=90)
+
+```
+
+```python
+plotlistofTuples(sorted(resultDictionary.items(), key=lambda tup: tup[0]), "Pings count per day")  
+```
+
+
+![png](images/output_23_0.png)
+
+
+
+```python
+pingsPerDaySeries = pd.Series(resultDictionary.values())
+pingsPerDaySeries.describe([.25, .5, .75, .95])
+```
+
+
+
+
+    count        27.000000
+    mean      74711.481481
+    std       37555.797027
+    min        6475.000000
+    25%       54222.500000
+    50%       70895.000000
+    75%       82903.000000
+    95%      138705.400000
+    max      164572.000000
+    dtype: float64
+
+
+
+# Nightly 56 & 57. # Compute how many clients are reporting
+
+
+```python
+def getClients(pings):
+    clients = pings.map(lambda ping: ping["clientId"]).distinct()
+    return clients
+```
+
+```python
+clientsCount = getClients(cachedData).count()
+print "Clients number = " + str(clientsCount)
+```
+    Clients number = 120336
+
+
+# Nightly 56 & 57. Compute average number of pings per client per day
+
+* Expected ping per day = 24, as we send no one Health ping per hour. 
+* This include showing diagrams and printings stats
+
+
+
+
+```python
+from collections import Counter
+
+def getAvgPerDate(iterable):
+    aggregare = Counter([d for d, r in iterable])
+    result = sum(aggregare.values()) * 1.0 / len(aggregare)
+    return result
+
+def pingsPerClientPerDay(pings, date):
+    return pings.map(lambda ping: (ping["clientId"], (ping[date], ping["payload/reason"]))).groupByKey()
+
+def avgPingsPerClientPerDay(pings, date):
+    idDateRDD = pingsPerClientPerDay(pings, date)
+    return idDateRDD.map(lambda pair: (pair[0], getAvgPerDate(pair[1]))).collect()
+```
+
+```python
+def plotAvgPingPerDateDistr(pings, date):
+    PINGS_COUNT_PER_DAY = 24
+    resultDistributionList = avgPingsPerClientPerDay(pings, date)
+    values = [v for k, v in resultDistributionList]
+    plt.title(date + ": clients sending more than expected ping per day - " + \
+                  str(len([v for v in values if v > PINGS_COUNT_PER_DAY])) + \
+                      ". Log on Y axis")
```

Comment:
If the y-axis is log, you should label the y-axis with "Ping counts (log)" instead of putting it in the title.

Actually, it's a good practice to label all of your axes in every plot.